File: go/muse/algorithms/sorting/quick_sort/quick_sort.go
package quick_sort

import "math/rand"

func Partition(arr []int, lo, hi int) int {
	pivot := arr[lo+((hi-1-lo)>>1)]

	i := lo - 1
	j := hi

	for {
		for {
			i++
			if arr[i] >= pivot {
				break
			}
		}
		for {
			j--
			if arr[j] <= pivot {
				break
			}
		}
		if i >= j {
			return j
		}
		arr[i], arr[j] = arr[j], arr[i]
	}
}

func Sort(arr []int) {
	SortWithRange(arr, 0, len(arr))
}

func SortWithRange(arr []int, lo, hi int) {
	if lo+1 >= hi {
		return
	}

	p := Partition(arr, lo, hi)

	SortWithRange(arr, lo, p+1)
	SortWithRange(arr, p+1, hi)
}

func PartitionThreeWay(arr []int, lo, hi int) (int, int) {
	i := lo - 1
	j := hi - 1

	p := lo - 1
	q := hi - 1

	pivot := arr[hi-1]

	for {
		for {
			i++
			if arr[i] >= pivot {
				break
			}
		}
		for {
			j--
			if pivot >= arr[j] {
				break
			}
			if j == lo {
				break
			}
		}
		if i >= j {
			break
		}
		arr[i], arr[j] = arr[j], arr[i]
		if arr[i] == pivot {
			p++
			arr[p], arr[i] = arr[i], arr[p]
		}
		if pivot == arr[j] {
			q--
			arr[j], arr[q] = arr[q], arr[j]
		}
	}

	arr[i], arr[hi-1] = arr[hi-1], arr[i]

	j = i - 1
	i++

	for k := lo; k < p; k++ {
		arr[k], arr[j] = arr[j], arr[k]
		j--
	}
	for k := hi - 2; k > q; k-- {
		arr[i], arr[k] = arr[k], arr[i]
		i++
	}

	return i, j
}

func SortThreeWayPartitionWithShuffle(arr []int) {
	rand.Shuffle(len(arr), func(i, j int) {
		arr[i], arr[j] = arr[j], arr[i]
	})
	SortWithRangeThreeWayPartition(arr, 0, len(arr))
}

func SortThreeWayPartition(arr []int) {
	SortWithRangeThreeWayPartition(arr, 0, len(arr))
}

func SortWithRangeThreeWayPartition(arr []int, lo, hi int) {
	if lo+1 >= hi {
		return
	}

	i, j := PartitionThreeWay(arr, lo, hi)

	SortWithRangeThreeWayPartition(arr, lo, j+1)
	SortWithRangeThreeWayPartition(arr, i, hi)
}

func PartitionInefficient(arr []int, lo, hi int) int {
	pivot := arr[lo]

	i := lo
	j := hi - 1

	for i != j {
		for j > i {
			if arr[j] < pivot {
				arr[i], arr[j] = arr[j], pivot
				break
			}
			j--
		}
		for i < j {
			if arr[i] > pivot {
				arr[j], arr[i] = arr[i], pivot
				break
			}
			i++
		}
	}

	return i
}

func SortInefficient(arr []int) {
	SortWithRangeInefficient(arr, 0, len(arr))
}

func SortWithRangeInefficient(arr []int, lo, hi int) {
	if lo >= hi {
		return
	}

	p := PartitionInefficient(arr, lo, hi)

	SortWithRangeInefficient(arr, lo, p)
	SortWithRangeInefficient(arr, p+1, hi)
}

File: rust/muse/src/algorithms/sorting/quick_sort.rs
use rand::{seq::SliceRandom, thread_rng};

pub fn partition(arr: &mut [i32], lo: usize, hi: usize) -> usize {
    let pivot = arr[lo + ((hi - 1 - lo) >> 1)];

    let mut i = lo.wrapping_sub(1);
    let mut j = hi;

    loop {
        loop {
            i = i.wrapping_add(1);
            if arr[i] >= pivot {
                break;
            }
        }
        loop {
            j -= 1;
            if arr[j] <= pivot {
                break;
            }
        }
        if i >= j {
            return j;
        }
        arr.swap(i, j);
    }
}

pub fn sort(arr: &mut [i32]) {
    sort_with_range(arr, 0, arr.len())
}

pub fn sort_with_range(arr: &mut [i32], lo: usize, hi: usize) {
    if lo + 1 >= hi {
        return;
    }

    let p = partition(arr, lo, hi);

    sort_with_range(arr, lo, p + 1);
    sort_with_range(arr, p + 1, hi);
}

pub fn partition_three_way(arr: &mut [i32], lo: usize, hi: usize) -> (usize, usize) {
    let mut i = lo.wrapping_sub(1);
    let mut j = hi - 1;

    let mut p = lo.wrapping_sub(1);
    let mut q = hi - 1;

    let pivot = arr[hi - 1];

    loop {
        loop {
            i = i.wrapping_add(1);
            if arr[i] >= pivot {
                break;
            }
        }
        loop {
            j -= 1;
            if pivot >= arr[j] {
                break;
            }
            if j == lo {
                break;
            }
        }
        if i >= j {
            break;
        }
        arr.swap(i, j);
        if arr[i] == pivot {
            p = p.wrapping_add(1);
            arr.swap(p, i);
        }
        if pivot == arr[j] {
            q -= 1;
            arr.swap(j, q);
        }
    }

    arr.swap(i, hi - 1);

    j = i.wrapping_sub(1);
    i += 1;

    let mut k = lo;
    while k < p && p != usize::MAX {
        arr.swap(k, j);
        j -= 1;
        k += 1;
    }
    k = hi - 2;
    while k > q {
        arr.swap(i, k);
        i += 1;
        k -= 1;
    }

    (i, j)
}

pub fn sort_three_way_partition_with_shuffle(arr: &mut [i32]) {
    let mut rng = thread_rng();
    arr.shuffle(&mut rng);
    sort_with_range_three_way_partition(arr, 0, arr.len())
}

pub fn sort_three_way_partition(arr: &mut [i32]) {
    sort_with_range_three_way_partition(arr, 0, arr.len())
}

pub fn sort_with_range_three_way_partition(arr: &mut [i32], lo: usize, hi: usize) {
    if lo + 1 >= hi {
        return;
    }

    let (i, j) = partition_three_way(arr, lo, hi);

    sort_with_range_three_way_partition(arr, lo, j.wrapping_add(1));
    sort_with_range_three_way_partition(arr, i, hi);
}

pub fn partition_inefficient(arr: &mut [i32], lo: usize, hi: usize) -> usize {
    let pivot = arr[lo];

    let mut i = lo;
    let mut j = hi - 1;

    while i != j {
        while j > i {
            if arr[j] < pivot {
                (arr[i], arr[j]) = (arr[j], pivot);
                break;
            }
            j -= 1;
        }
        while i < j {
            if arr[i] > pivot {
                (arr[j], arr[i]) = (arr[i], pivot);
                break;
            }
            i += 1;
        }
    }

    i
}

pub fn sort_inefficient(arr: &mut [i32]) {
    sort_with_range_inefficient(arr, 0, arr.len())
}

pub fn sort_with_range_inefficient(arr: &mut [i32], lo: usize, hi: usize) {
    if lo >= hi {
        return;
    }

    let p = partition_inefficient(arr, lo, hi);

    sort_with_range_inefficient(arr, lo, p);
    sort_with_range_inefficient(arr, p + 1, hi);
}

#[cfg(test)]
mod tests;
